tests/unit/test_app_endpoints.py

```diff
@@ -0,0 +1,122 @@
+"""Unit tests for app endpoints."""
+
+import os
+from http import HTTPStatus
+from unittest.mock import patch
+
+import pytest
+from fastapi import HTTPException
+
+from ols import constants
+from ols.app.endpoints import feedback, ols
+from ols.app.models.models import FeedbackRequest, LLMRequest
+from ols.utils import config
+
+
+@pytest.fixture(scope="module")
+@patch.dict(os.environ, {"LOG_LEVEL": "INFO"})
+@patch.dict(os.environ, {"BAM_API_KEY": "BAM API KEY"})
+def load_config():
+    """Load config before unit tests."""
+    config.load_config_from_env()
+
+
+def test_feedback_request(load_config):
+    """Test feedback request."""
+    feedback_request = FeedbackRequest(
+        conversation_id=123,
+        feedback_object='{"rating": 5, "comment": "Great service!"}',
+    )
+    response = feedback.feedback_request(feedback_request)
+    assert response == {"status": "feedback received"}
+
+
+@patch("ols.src.query_helpers.question_validator.QuestionValidator.validate_question")
+@patch("ols.src.docs.docs_summarizer.DocsSummarizer.summarize")
+@patch("ols.src.query_helpers.yaml_generator.YamlGenerator.generate_yaml")
+@patch("ols.utils.config.conversation_cache.get")
+def test_ols_request(
+    mock_conversation_cache_get,
+    mock_generate_yaml,
+    mock_summarize,
+    mock_validate_question,
```

Comment:
is there some python magic that is mapping the mocks created by the patch commands to these argument names?  I don't doubt it works but i'd like to understand how.

Reply:
These mock functions' references are added by the `@patch` decorator in a "first in last out" order. In the test we can use them to modify their behavior such as their return values for different test cases. 
For example we can let the function `ols.src.query_helpers.question_validator.QuestionValidator.validate_question` return `[constants.VALID, constants.NOYAML]` for the 1st call while return `[constants.VALID, constants.YAML]` for the 2nd call in this test.

Reply:
ok, so there's an ordering relationship between the @patch annotations and the args being passed to the test function.  But the order is inverted(first @patch annotation -> last argument).  Got it.

Reply:
yes. The fact is that any decorator __transforms__ the original function into some other function. So the "last" decorator takes the original function X, producing function Y. That function Y is basically passed into "last-1" decorator, producing function Z etc. That means - decorators are higher order function called during module import.

(that idea is super nice, so IMHO it's good to know what's going on - decorators are different beasts than, for example, annotations in Java, just by coincidence are written similarly)